cmd/liquidator.go

```diff
@@ -0,0 +1,138 @@
+package cmd
+
+import (
+	"context"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// LiquidationTarget contains the address of a borrower who is over their borrow limit, as
+// well as their current borrowed amounts and collateral in all denominations.
+// CONTRACT: Collateral uTokens must be expressed as their equivalent value in base tokens.
+type LiquidationTarget struct {
+	Addr       sdk.AccAddress
+	Borrowed   sdk.Coins
+	Collateral sdk.Coins
+}
+
+// LiquidationOrder expresses the intent to perform, or the outcome of, a liquidation.
+// Collateral reward amounts are always expressed in base tokens equivalent to their
+// underlying uToken amount.
+type LiquidationOrder struct {
+	Addr   sdk.AccAddress
+	Repay  sdk.Coin
+	Reward sdk.Coin
+}
+
+// startLiquidator causes the liquidator to continuously look for liquidation targets, decide on which
+// borrowed and collateral denominations to attempt to liquidate, and attempt to execute any
+// liquidations whose estimated outcomes are approved by its configured decisionmaking. Returns
+// only when context is cancelled.
+func startLiquidator(
+	ctx context.Context,
+	cancelFunc context.CancelFunc,
+) error {
+
+	// loop as long as ctx is not cancelled
+	for ctx.Err() == nil {
+		// get a list of eligible liquidation targets
+		targets, err := getLiquidationTargets(ctx)
+		if err != nil {
+			logger.Err(err)
+			continue
+		}
+
+		for _, target := range targets {
+			if ctx.Err() == nil {
+				// select one reward denom and one repay denom to consider on target address
+				intent, ok, err := selectLiquidationDenoms(ctx, target)
+				if err != nil {
+					logger.Err(err)
```

Comment:
any log that we log with an error, we should also accompany a `Msg(...)` for additional context :)

Reply:
done at [9c1181c](https://github.com/umee-network/liquidator/pull/1/commits/9c1181cebb22a7ed2c7bf78e2d24f1d5e8b008fb)